[FEATURE] Adding True After middlewares to ExpressJS framework.

ganeshkbhat opened the pull request:
Adding `True After middlewares` to ExpressJS framework.

1. The following commit do not change anything in the code logic but `impacts the code request-response lifcycle/ flow`.
2. The arguments of the `res.send` function that takes only response body as its arguments.
3. A callback called `afterCallback` has been added that modifies the function to take a callback that can be executed after a response has been sent.
4. The previous code lifecycle/ flow was `Request -> Before Middlewares -> APILogic -> Response`
5. The new code lifecycle/ flow will `Request -> Before Middlewares -> APILogic -> Response -> After Middleware`
6. NOTE: The code does not allow for mutliple after callbacks like the middlewares. It adds only one callback function to avoid complexity of future changes and testing issues to the framework. One function called afterCallback is a developer tradeoff to manage and test their own code base for the after middleware (afterCallback) that can combine multiple after middlewares into one after middleware.

Pull requests (Changes mentioned below):
FILENAME: [https://github.com/expressjs/express/blob/master/lib/response.js](https://github.com/expressjs/express/blob/master/lib/response.js)

Code Changes below:

[A]:
```
res.send(body) => res.send(body, afterCallback, ...args)
```
      Modified Lines:
      1. Line 113: `res.send = function send(body, afterCallback, ...args) {`
      2. Line 237: `// True after middlewares`
      3. Line 238: `if (!afterCallback || afterCallback !== "function") { afterCallback = (r, ...a) => {} }`
      4. Line 239: `((afterCallback, response, ...args) => { afterCallback(response, ...args); })(afterCallback, this, ...args);`

[B]:
```
res.json(obj) => res.json(obj, afterCallback, ...args)
```
	Modified Lines:
	1. Line 302: `res.json = function json(obj, afterCallback, ...args) {`
	2. Line 285: `return this.send(body, afterCallback, ...args);`

[C]:
```
res.jsonp(obj) => res.jsonp(obj, afterCallback, ...args)
```
	Modified Lines:
	1. Line 302: `res.jsonp = function jsonp(obj, afterCallback, ...args) {`
	2. Line 360: `return this.send(body, afterCallback, ...args);`

Commit:
Adding After middlewares [FEATURE] to ExpressJS framework.

1. The following commit do not change anything in the code logic but impacts the code flow.
2. The arguments of the res.send function that takes only response body as its arguments.
3. A callback called afterCallback has been added that modifies the function to take a callback that can be executed after a response has been sent.
4. The previous code flow was Request -> Before Middlewares -> APILogic -> Response
5. The new code flow will Request -> Before Middlewares -> APILogic -> Response -> After Middleware
6. NOTE: The code does not allow for mutliple callbacks. It adds only one callback function to avoid complexity of future changes and testing issues to the framework.
7. One function called afterCallback is a developer tradeoff to manage and test their own code base for the after middleware (afterCallback) that can combine multiple after middlewares into one after middleware.

Pull requests (Changes mentioned below):

res.send(body) => res.send(body, afterCallback, ...args)
	Modified Lines:
	1. Line 113: res.send = function send(body, afterCallback, ...args) {
	2. Line 237: // True after middlewares
	3. Line 238: ((afterCallback, response, ...args) => { afterCallback(response, ...args); })(afterCallback, this, ...args);

res.json(obj) => res.json(obj, afterCallback, ...args)
	Modified Lines:
	1. Line 302: res.json = function json(obj, afterCallback, ...args) {
	2. Line 285: return this.send(body, afterCallback, ...args);

res.jsonp(obj) => res.jsonp(obj, afterCallback, ...args)
	Modified Lines:
	1. Line 302: res.jsonp = function jsonp(obj, afterCallback, ...args) {
	2. Line 360: return this.send(body, afterCallback, ...args);

dougwilson commented:
Hello, and thank you for your pull request. It seems what you adding here can already be accomplished, and more generally with support for `.end` and `.pipe` too, by adding a listener to the `'finish'` event, or even using the `stream.finished`, depending on what you are trying to accomplish.

I would provide an example, but you don't have an example for the use-cases this provides. If you don't think those events handle what you are trying to achieve, please, can you elaborate more on the specific use-case for this and what specifically it achieves over the existing events on `res`.

Thank you!

ganeshkbhat committed:
Updating code for handling callback function not defined error. Adding new line to line 238

Updating code for handling callback function not defined error. Adding new line to line 238.

Modified line for [A] now:

[A]:
`res.send(body) => res.send(body, afterCallback, ...args)`

Modified Lines:
1. Line 113: `res.send = function send(body, afterCallback, ...args) {`
2. Line 237: `// True after middlewares`
3. Line 238: `if (!afterCallback || afterCallback !== "function") { afterCallback = (r, ...a) => {} }`
4. Line 239: `((afterCallback, response, ...args) => { afterCallback(response, ...args); })(afterCallback, this, ...args);`

ganeshkbhat committed:
Updating code for handling callback function not defined error. Adding new line to line 238

Updating code for handling callback function not defined error. Adding new line to line 238

Updating code for handling callback function not defined error. Missed typeof assertion Adding new line to line 238.

Modified line for [A] now:

[A]:
`res.send(body) => res.send(body, afterCallback, ...args)`

Modified Lines:
1. Line 113: `res.send = function send(body, afterCallback, ...args) {`
2. Line 237: `// True after middlewares`
3. Line 238: `if (!afterCallback || afterCallback !== "function") { afterCallback = (r, ...a) => {} }`
4. Line 239: `((afterCallback, response, ...args) => { afterCallback(response, ...args); })(afterCallback, this, ...args);`

ganeshkbhat commented:
Updating code to handle errors if `afterCallback` is not defined.

        Modified Lines:
        1. Line 113: `res.send = function send(body, afterCallback, ...args) {`
        2. Line 237: `// True after middlewares`
        3. Line 238: `if (!afterCallback || afterCallback !== "function") { afterCallback = (r, ...a) => {} }`
        4. Line 239: `((afterCallback, response, ...args) => { afterCallback(response, ...args); })(afterCallback, this, ...args);`

ganeshkbhat commented:
True @dougwilson dougwilson I am also considering a new pull request as well. Seems like there are three/ four code patterns we can use:


1. [CURRENT PULL REQUEST]: Response `res.send`, `res.json`, `res.jsonp` Defintion transformed into `Request -> Before Middlewares -> APILogic -> Response -> After Middleware`
    * `res.send(body)` to `res.send(body, afterCallback, ...args)`
        * <sub>body: <string|number|boolean|object|Buffer></sub>, <sub>afterCallback: <function|function[]|undefined></sub>
    * `res.json(obj)` to `res.json(obj, afterCallback, ...args)`
        * <sub>obj: <string|number|boolean|object></sub>, <sub>afterCallback: <function|function[]|undefined></sub>
    * `res.jsonp(obj)` to `res.jsonp(obj, afterCallback, ...args)`
        * <sub>obj: <string|number|boolean|object></sub>, <sub>afterCallback: <function|function[]|undefined></sub>
    * The new pull request after handling undefined afterCallback definitions makes this pull request to be backward compatible for the functions to be used as `res.send(body)`, `res.json(obj)`, `res.jsonp(obj)`.


2. [CONSIDERING NEW PULL REQUEST]: Route Defintion transformed from `express[method<get>](path, middlewares, callback)` to `express[method<get>](path, middlewares, callback, afterMiddlewares)`:
    * From `express.get('/', middlewares, callback)` into `express.get('/', middlewares, callback, afterCallback)`.
        * <sub>method<get|put|patch|delete></sub>, <sub>middlewares: <function|function[]|undefined></sub>, <sub>callback: <function.></sub>, <sub>afterMiddlewares: <function|function[]|undefined></sub>


3. [SUGGESTED]: Route using `finish` event `addeventlistener` `finish` event of `stream.finish`. I believe you are refering to this [event finish feature](https://stackoverflow.com/questions/6537510/node-js-https-no-response-end-event-close-instead)

Can you elaborate on usage for `.end` or `.pipe` usage <sub>[SUGGESTED]</sub>? Did you mean `.pipe` to the `finish` listener.

I have made a small change to the codebase to handle `afterCallback` not defined by user. It will be upto the user to defined the after callback or not. The use cases for a `res.send(body)` or `express.post(path, callback)` based afterCallback after response middlewares may be: synchronous execution of `cleanup activities`, `needing same route based authentication, same sessions activities, etc`; which are devops-like use case in nature.

Note 1: Testing difficulties, or pros, or cons will be the same as for afterCallback based after response middleware or for after response finish event handling; while I believe at this points 2, and 3 are more testable.

Note 2: I also recommend a change in `afterCallback` naming to probably `callback` or something else. But I leave the choice to the team `if` the after response callback feature is accepted.

ganeshkbhat committed:
Update backward support oldest node v0.0.10, v0.0.12 for spread version

Update backward commit to support oldest node v0.0.10, v0.0.12 - [checking tests - build failing]

1. Managing minor changes to arguments passing to remove spread operator which is not supported by oldest node version 0.0.10.
2. Minor changes to afterCallback invocation. Inverting falsy check to a truthy check invocation for undefined checks for `res.send(body)` usage.
3. Using flatten function to flatten arguments passed.
4. Renaming afterCallback to afterSend
5. Test for `1) app.router:  should be .use()able: Error: expected 200 "OK", got 500 "Internal Server Error"` is expected to fail for res.send(body, afterSend). Do you need me to write tests for the same or the team will review the tests cases considering other tests?

`1) app.router:  should be .use()able: Error: expected 200 "OK", got 500 "Internal Server Error"`

```
Error: expected 200 "OK", got 500 "Internal Server Error"
      at Context.<anonymous> (test/app.router.js:161:6)
      at processImmediate (node:internal/timers:464:21)
```

ganeshkbhat commented:
**Update [3] for build and failing tests.**
`update/ commit for backward compatibility to support oldest v0.0.10, v0.0.12 - [checking tests - build failing]`

1. Managing minor changes to arguments passing to remove spread operator `...args` usage which is not supported by oldest node version 0.0.10.
2. Minor changes to afterCallback invocation. Inverting falsy check to a truthy check invocation for undefined checks for `res.send(body)` usage.
3. Using flatten function to flatten arguments passed.
4. Renaming afterCallback to afterSend
5. Test for `1) app.router:  should be .use()able: Error: expected 200 "OK", got 500 "Internal Server Error"` is expected to fail for res.send(body, afterSend). 

@dougwilson Failing tests. `1) app.router:  should be .use()able: Error: expected 200 "OK", got 500 "Internal Server Error"`
```
Error: expected 200 "OK", got 500 "Internal Server Error"
      at Context.<anonymous> (test/app.router.js:161:6)
      at processImmediate (node:internal/timers:464:21)
```

**Note:** @dougwilson Do you need me to write tests for the same or the team will review the tests cases considering other tests and use cases? The `app.router:  should be .use()able:` has to be rewritten to accomodate new feature.

dougwilson commented:
Hi @ganeshkbhat I would say before you get too far, I am still unclear on what exact use-case this enables that is not already possible. Just looking at the code, it doesn't seem like it just takes a function and calls it right before the `return` in `res.send`, which just adds what seems like a lot of complication that would just be identical to simply calling that function after making the `res.send` call in your handler. Maybe it would help if you could provide a specific code example that would make use of this feature that is not already possible with how Express.js already works.

ganeshkbhat commented:
> Hi @ganeshkbhat I would say before you get too far, I am still unclear on what exact use-case this enables that is not already possible. Just looking at the code, it doesn't seem like it just takes a function and calls it right before the `return` in `res.send`, which just adds what seems like a lot of complication that would just be identical to simply calling that function after making the `res.send` call in your handler. Maybe it would help if you could provide a specific code example that would make use of this feature that is not already possible with how Express.js already works.

@dougwilson for the complexity which even I was considering I am considering alternate architecture pull request.

* [CONSIDERING NEW PULL REQUEST]
     - Route Defintion transformed from `express[method<get>](path, middlewares, callback)` to `express[method<get>](path, middlewares, callback, afterSend)`:
     - From `express.get('/', middlewares, callback)` into `express.get('/', middlewares, callback, afterSend)`.
     - method: `<get|post|put|patch|delete|options>`
     - middlewares: `<function|function[]|undefined>`
     - callback: `<function.>`
     - afterSend: `<function|function[]|undefined>`


@dougwilson [Gulp](https://gulpjs.com/) or [Jenkins pipeline like](https://www.jenkins.io/doc/book/pipeline/running-pipelines/) DevOps `HTTP/S REST - GraphQL based Granular Authenticated and Authorized Server`.

Here is a demo task

```
const express = require('express');
const app = express();

app.use(function (req, res, next) {
    next();
});

app.get("/task", function (req, res, next) {
    res.status(200).send("Send:get:/test: Testing after Middlewares " + Date.now().toString(), (response) => { console.log("DevOps Task" + Date.now().toString()); });
});

app.get("/", (req, res, next) => { console.log("middleware invoked " + Date.now().toString()); next(); }, function (req, res, next) {
    next();
});

app.use(function (req, res, next) {
    res.status(200).send("Send:get:/: Testing after Middlewares " + Date.now().toString(), (response) => { console.log("HTTP/S DevOps/ Queue cleanup same auth and same session task " + Date.now().toString()); });
});

app.listen(3000, () => { console.log("Starting DevOps HTTP/S REST - GraphQL - gRPC Server"); });

```

dougwilson commented:
Thank you for that. It seems when running that, there is no difference from the following code, however:
```js
const express = require('express');
const app = express();

app.use(function (req, res, next) {
    next();
});

app.get("/task", function (req, res, next) {
    res.status(200).send("Send:get:/test: Testing after Middlewares " + Date.now().toString());
    console.log("DevOps Task" + Date.now().toString());
});

app.get("/", (req, res, next) => { console.log("middleware invoked " + Date.now().toString()); next(); }, function (req, res, next) {
    next();
});

app.use(function (req, res, next) {
    res.status(200).send("Send:get:/: Testing after Middlewares " + Date.now().toString());
    console.log("HTTP/S DevOps/ Queue cleanup same auth and same session task " + Date.now().toString());
});

app.listen(3000, () => { console.log("Starting DevOps HTTP/S REST - GraphQL - gRPC Server"); });
```

Please explain what I am missing here, as I do not understand what this pull request is doing that is different from the above.

ganeshkbhat commented:
@dougwilson Absolutely True. My bad. I am still wondering why this was never uncovered by me ever over the last few years.


**[1].**
@dougwilson Was this since long time before or just recently added. I always remembered it failed such implementation and more so while setting headers after a `res.send`. I always and never ran a function after the `res.send` or `res.json`, `res.jsonp`, `res.render`.

I have been searching the whole internet and I am at aghast why this architecture of running a task after sending a response is scarcely available; rather wish to say not available other than point two like link below: `Request -> Before Middlewares -> APILogic -> Response -> After Middleware`

Look at some sample codes and pointers below:

1. [Execute a function after ExpressJS res.send https://stackoverflow.com/questions/28868842/expressjs-4-execute-function-after-res-send](https://stackoverflow.com/questions/28868842/expressjs-4-execute-function-after-res-send)
2. [Execute a piping of promises after ExpressJS res.send https://gist.github.com/bennadel/c8646e75d1e8fb3bd3d0e15e53167b3a](https://gist.github.com/bennadel/c8646e75d1e8fb3bd3d0e15e53167b3a)
3. [Multiple HTTP writes before reponse.end https://blog.kevinchisholm.com/javascript/node-js/express-js/response-send-end-write-difference/](https://blog.kevinchisholm.com/javascript/node-js/express-js/response-send-end-write-difference/)
4. [ExpressJS finish response event https://stackoverflow.com/questions/12239731/node-js-express-on-response-event](https://stackoverflow.com/questions/12239731/node-js-express-on-response-event)


**[2].**
@dougwilson I was considering how I would or even why I would you use a `.end` and `.pipe` if functions can run after the `res.send` function can be invoked. Well, atleast why use .pipe if not .end since the use cases are entirely different.


**[3].**
@dougwilson To be frank `https`'s `response.end` event is far different from the response `close`/ response `finish` is far different from running a after send middleware or rather a group of middlewares after send.

```
Hello, and thank you for your pull request. It seems what you adding here can already be accomplished, and more generally with support for .end and .pipe too, by adding a listener to the 'finish' event, or even using the stream.finished, depending on what you are trying to accomplish.

I would provide an example, but you don't have an example for the use-cases this provides. If you don't think those events handle what you are trying to achieve, please, can you elaborate more on the specific use-case for this and what specifically it achieves over the existing events on res.
```


**[4].**
But, well, yes, you can always apply a middleware like below for all of response's `finish` event; and it should work.

```
app.use(function(req, res,next){
    res.on('finish', function(){
        console.log('Finish<Event>: the response has been sent. Executing DevSecOps task');
    });
    next();
});
```


**Pros:**
- Apply middleware for all requests
- Apply after send or finish event for all request as a middleware

**Cons:**
- @dougwilson  Authentication and Authorization complications for all routes [single or all method], specific route [single or all methods]
- @dougwilson True after send middlewares like before response sent middlewares usage
- @dougwilson Specific route and method applicability for all routes [single or all method], specific route [single or all methods]


**[5].**
@dougwilson I am considering this architecture below in point 6 for a new pull request considering that a function can be invoked after the response `send` event.

```
res.status(200).send("response sent")
console.log("DevSecOps After send function of any architecture or functionality")
```

**[6].**

@dougwilson Considering the proposal of - Implement the new architecture of `Request -> Before Middlewares -> APILogic -> Response -> After Middleware` that changes `express[method<get>](path, middlewares, callback)` to `express[method<get>](path, middlewares, callback, afterMiddlewares)`.

It will change current implementation of route from `express.get('/', middlewares, callback)` into `express.get('/', middlewares, callback, afterSendMiddlewares)`

     - method: `<get|post|put|patch|delete|options>`
     - middlewares: `<function|function[]|undefined>`
     - callback: `<function.>`
     - afterSendMiddlewares: `<function|function[]|undefined>`

_[CONSIDERATION] [PROPOSAL] [NEW FEATURE] [TESTS NEEDED] [NEW ARCHITECTURE]_

dougwilson commented:
Ok. I'm going to close this pull request for the time being since, as I pointed it, it doesn't actually add anything not already poasible. Regarding your other throughs, I encourage you to open a new issue to outline your thoughts and ideas, demonstrate your comcrete use cases they enable, and let folks hash it out prior to opening a pull request. This is particularly because your suggestion is going to change the key signatures of core functions, so we want to work out some of those details in an issue ideally 👍

ganeshkbhat commented:
@dougwilson True. Should I send the new proposal pull request. I am seriously considering this. I am in a consideration for the pull request `not to touch any <T> or methods` but `just` the `internals and explicit arguments extensions (including version backward compatibility)` which will have newer isolated implementations to current implementations. I did not and may not be able to figure out key signatures yet. I may have issues

---------------------------------

_[CONSIDERATION] [PROPOSAL] [NEW FEATURE] [TESTS NEEDED] [NEW ARCHITECTURE]_

---------------------------------

Implement the new architecture of `Request -> Before Middlewares -> APILogic -> Response -> After Middleware` that changes `express[method<get>](path, middlewares, callback)` to `express[method<get>](path, middlewares, callback, afterMiddlewares)`.

It will change current implementation of route from `express.get('/', middlewares, callback)` into `express.get('/', middlewares, callback, afterSendMiddlewares)`

     - method: `<get|post|put|patch|delete|options>`
     - middlewares: `<function|function[]|undefined>`
     - callback: `<function.>`
     - afterSendMiddlewares: `<function|function[]|undefined>`

* Authentication and Authorization complications removal for all routes [single or all method], specific route [single or all methods]
* True after send middlewares like before response sent middlewares usage
* Specific route and method applicability for all routes [single or all method], specific route [single or all methods]

This pull request may need the many tests to be changed for `routes` and route `Layer` methods

---------------------------------

@dougwilson Can you add any views here with your immense experience on the internals of expressjs? I would love to keep that in the back of my mind. You probably saw what I goal I am trying to reach.
